File: 2024/AoC24-08ab.rs
use std::fs;
use std::collections::{HashSet, HashMap};
use std::ops::{Add, Sub, AddAssign, SubAssign};

fn main() {
  let input = fs::read_to_string("input.txt").unwrap();

  let antennae = process(&input);
  let bounds = {
    let width = input.lines().next().unwrap().chars().count() as i32;
    let height = input.lines().count() as i32;

    Point {
      x: width,
      y: height,
    }
  };
  
  println!("Part 1 answer: {}", part1(&antennae, bounds));
  println!("Part 2 answer: {}", part2(&antennae, bounds));
}

fn part1(antennae: &HashMap<char, Vec<Point>>, bounds: Point) -> usize {
  let mut antinodes: HashSet<Point> = HashSet::new();

  for (_, val) in antennae {
    for i in 0..(val.len()-1) {
      for j in (i+1)..val.len() {
        let (a, b) = (val[i], val[j]);

        let diff = a - b;
        antinodes.insert(a + diff);
        antinodes.insert(b - diff);
      }
    }
  }

  antinodes.into_iter().filter(|p| p.is_within(&bounds)).count()
}

fn part2(antennae: &HashMap<char, Vec<Point>>, bounds: Point) -> usize {
  let mut antinodes: HashSet<Point> = HashSet::new();

  for (_, val) in antennae {
    for i in 0..(val.len()-1) {
      for j in (i+1)..val.len() {
        let (mut a, mut b) = (val[i], val[j]);

        let diff = a - b;
        while a.is_within(&bounds) {
          antinodes.insert(a);
          a += diff;
        }
        while b.is_within(&bounds) {
          antinodes.insert(b);
          b -= diff;
        }
      }
    }
  }

  antinodes.len()
}

fn process(input: &str) -> HashMap<char, Vec<Point>> {
  let mut antennae: HashMap<char, Vec<Point>> = HashMap::new();
  for (y, line) in input.lines().enumerate() {
    for (x, c) in line.chars().enumerate() {
      if c != '.' {
        antennae.entry(c).or_insert(vec![]).push(Point { x: x as i32, y: y as i32 });
      }
    }
  }

  antennae
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
struct Point {
  x: i32,
  y: i32,
}

impl Point {
  fn is_within(&self, bounds: &Self) -> bool {
    (0..bounds.x).contains(&self.x) && (0..bounds.y).contains(&self.y)
  }
}

impl Add for Point {
  type Output = Self;
  fn add(self, other: Self) -> Self {
    Point {
      x: self.x + other.x,
      y: self.y + other.y,
    }
  }
}

impl AddAssign for Point {
  fn add_assign(&mut self, other: Self) {
    *self = *self + other;
  }
}

impl Sub for Point {
  type Output = Self;
  fn sub(self, other: Self) -> Self {
    Point {
      x: self.x - other.x,
      y: self.y - other.y,
    }
  }
}

impl SubAssign for Point {
  fn sub_assign(&mut self, other: Self) {
    *self = *self - other;
  }
}

File: 2024/AoC24-08ab.dart
import 'dart:io';
import 'dart:core';
import 'dart:collection';

void main() {
  final String input = File('input.txt').readAsStringSync();

  final (antennae, bounds) = process(input);

  print('Part 1 answer: ${part1(antennae, bounds)}');
  print('Part 2 answer: ${part2(antennae, bounds)}');
}

int part1(List<List<Point>> antennae, Point bounds) {
  Set<Point> antinodes = HashSet<Point>();

  for (final group in antennae) {
    for (final (a, b) in group.unorderedPairs()) {
      final Point diff = a - b;
      antinodes.add(a + diff);
      antinodes.add(b - diff);
    }
  }

  return antinodes.where((p) => p.isWithin(bounds)).length;
}

int part2(List<List<Point>> antennae, Point bounds) {
  Set<Point> record = HashSet<Point>();

  for (final group in antennae) {
    for (var (a, b) in group.unorderedPairs()) {
      final Point diff = a - b;
      while (a.isWithin(bounds)) {
        record.add(a);
        a += diff;
      }
      while (b.isWithin(bounds)) {
        record.add(b);
        b -= diff;
      }
    }
  }

  return record.length;
}

(List<List<Point>>, Point) process(String input) {
  final List<List<String>> grid = input.split('\n').map((line) => line.split('').toList()).toList();

  Map<String, List<Point>> antennae = Map();
  for (final (i, row) in grid.enumerate()) {
    for (final (j, c) in row.enumerate()) {
      if (c == '.') continue;
      List<Point> v = antennae[c] ?? [];
      v.add(Point(j, i));
      antennae[c] = v;
    }
  }

  final int width = grid[0].length;
  final int height = grid.length;
  final Point bounds = Point(width, height);

  return (antennae.values.toList(), bounds);
}

class Point {
  final int x, y;

  Point(this.x, this.y);
  Point.clone(Point p) : this(p.x, p.y);

  bool isWithin(Point bounds) =>
    0 <= this.x && this.x < bounds.x && 0 <= this.y && this.y < bounds.y;

  Point operator +(Point other) =>
    Point(this.x + other.x, this.y + other.y);

  Point operator -(Point other) =>
    Point(this.x - other.x, this.y - other.y);

  @override
  bool operator ==(Object o) =>
    o is Point && this.x == o.x && this.y == o.y;

  @override
  int get hashCode => Object.hash(x, y);
}

extension <E> on List<E> {
  List<(int, E)> enumerate([start = 0]) {
    List<(int, E)> out = [];
    for (final E e in this) out.add((start++, e));

    return out;
  }

  List<(E, E)> unorderedPairs() {
    List<(E, E)> out = [];
    for (final (i, e) in this.enumerate().sublist(1)) {
      out.addAll(this.sublist(0, i).map((f) => (e, f)));
    }

    return out;
  }
}